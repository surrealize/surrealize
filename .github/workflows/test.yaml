name: Test & Build

on:
  push:
    branches:
      - "**"
      - "!next"
      - "!main"
  workflow_call:

jobs:
  build:
    name: Test & Build
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: oven-sh/setup-bun@v2
        with:
          bun-version: latest
      - uses: surrealdb/setup-surreal@v2
        with:
          surrealdb_auth: false
      - run: bun install --frozen-lockfile
      - run: bun run check
      - run: bun test
      - run: bun run build
