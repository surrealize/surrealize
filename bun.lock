{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "dependencies": {
        "@standard-schema/spec": "^1.0.0",
        "cbor2": "^1.12.0",
      },
      "devDependencies": {
        "@std/semver": "npm:@jsr/std__semver@^1.0.4",
        "@trivago/prettier-plugin-sort-imports": "^5.2.2",
        "@types/bun": "^1.2.4",
        "arktype": "^2.1.2",
        "dayjs": "^1.11.13",
        "dts-bundle-generator": "^9.5.1",
        "prettier": "^3.5.3",
        "typescript": "^5.8.2",
      },
    },
  },
  "packages": {
    "@ark/schema": ["@ark/schema@0.43.1", "", { "dependencies": { "@ark/util": "0.43.1" } }, "sha512-B28ceusolthB9bmHQEG9Lp16keUTt5JhuoB9Uhaw7MQh9+ODSbH4d2KJDD/yrMSeoXiNqEGuWmhlR6YO2pdNvA=="],

    "@ark/util": ["@ark/util@0.43.1", "", {}, "sha512-sRx5bZiNoilA7mr5lAu78ZHOJx88nCagLER9Ns1FfoWWHrDWj8J8xU+VFj0g1ujJrAxWsVJRVElOEZ0XzXCrDw=="],

    "@babel/code-frame": ["@babel/code-frame@7.26.2", "", { "dependencies": { "@babel/helper-validator-identifier": "^7.25.9", "js-tokens": "^4.0.0", "picocolors": "^1.0.0" } }, "sha512-RJlIHRueQgwWitWgF8OdFYGZX328Ax5BCemNGlqHfplnRT9ESi8JkFlvaVYbS+UubVY6dpv87Fs2u5M29iNFVQ=="],

    "@babel/generator": ["@babel/generator@7.26.8", "", { "dependencies": { "@babel/parser": "^7.26.8", "@babel/types": "^7.26.8", "@jridgewell/gen-mapping": "^0.3.5", "@jridgewell/trace-mapping": "^0.3.25", "jsesc": "^3.0.2" } }, "sha512-ef383X5++iZHWAXX0SXQR6ZyQhw/0KtTkrTz61WXRhFM6dhpHulO/RJz79L8S6ugZHJkOOkUrUdxgdF2YiPFnA=="],

    "@babel/helper-string-parser": ["@babel/helper-string-parser@7.25.9", "", {}, "sha512-4A/SCr/2KLd5jrtOMFzaKjVtAei3+2r/NChoBNoZ3EyP/+GlhoaEGoWOZUmFmoITP7zOJyHIMm+DYRd8o3PvHA=="],

    "@babel/helper-validator-identifier": ["@babel/helper-validator-identifier@7.25.9", "", {}, "sha512-Ed61U6XJc3CVRfkERJWDz4dJwKe7iLmmJsbOGu9wSloNSFttHV0I8g6UAgb7qnK5ly5bGLPd4oXZlxCdANBOWQ=="],

    "@babel/parser": ["@babel/parser@7.26.8", "", { "dependencies": { "@babel/types": "^7.26.8" }, "bin": "./bin/babel-parser.js" }, "sha512-TZIQ25pkSoaKEYYaHbbxkfL36GNsQ6iFiBbeuzAkLnXayKR1yP1zFe+NxuZWWsUyvt8icPU9CCq0sgWGXR1GEw=="],

    "@babel/template": ["@babel/template@7.26.8", "", { "dependencies": { "@babel/code-frame": "^7.26.2", "@babel/parser": "^7.26.8", "@babel/types": "^7.26.8" } }, "sha512-iNKaX3ZebKIsCvJ+0jd6embf+Aulaa3vNBqZ41kM7iTWjx5qzWKXGHiJUW3+nTpQ18SG11hdF8OAzKrpXkb96Q=="],

    "@babel/traverse": ["@babel/traverse@7.26.8", "", { "dependencies": { "@babel/code-frame": "^7.26.2", "@babel/generator": "^7.26.8", "@babel/parser": "^7.26.8", "@babel/template": "^7.26.8", "@babel/types": "^7.26.8", "debug": "^4.3.1", "globals": "^11.1.0" } }, "sha512-nic9tRkjYH0oB2dzr/JoGIm+4Q6SuYeLEiIiZDwBscRMYFJ+tMAz98fuel9ZnbXViA2I0HVSSRRK8DW5fjXStA=="],

    "@babel/types": ["@babel/types@7.26.8", "", { "dependencies": { "@babel/helper-string-parser": "^7.25.9", "@babel/helper-validator-identifier": "^7.25.9" } }, "sha512-eUuWapzEGWFEpHFxgEaBG8e3n6S8L3MSu0oda755rOfabWPnh0Our1AozNFVUxGFIhbKgd1ksprsoDGMinTOTA=="],

    "@jridgewell/gen-mapping": ["@jridgewell/gen-mapping@0.3.5", "", { "dependencies": { "@jridgewell/set-array": "^1.2.1", "@jridgewell/sourcemap-codec": "^1.4.10", "@jridgewell/trace-mapping": "^0.3.24" } }, "sha512-IzL8ZoEDIBRWEzlCcRhOaCupYyN5gdIK+Q6fbFdPDg6HqX6jpkItn7DFIpW9LQzXG6Df9sA7+OKnq0qlz/GaQg=="],

    "@jridgewell/resolve-uri": ["@jridgewell/resolve-uri@3.1.2", "", {}, "sha512-bRISgCIjP20/tbWSPWMEi54QVPRZExkuD9lJL+UIxUKtwVJA8wW1Trb1jMs1RFXo1CBTNZ/5hpC9QvmKWdopKw=="],

    "@jridgewell/set-array": ["@jridgewell/set-array@1.2.1", "", {}, "sha512-R8gLRTZeyp03ymzP/6Lil/28tGeGEzhx1q2k703KGWRAI1VdvPIXdG70VJc2pAMw3NA6JKL5hhFu1sJX0Mnn/A=="],

    "@jridgewell/sourcemap-codec": ["@jridgewell/sourcemap-codec@1.5.0", "", {}, "sha512-gv3ZRaISU3fjPAgNsriBRqGWQL6quFx04YMPW/zD8XMLsU32mhCCbfbO6KZFLjvYpCZ8zyDEgqsgf+PwPaM7GQ=="],

    "@jridgewell/trace-mapping": ["@jridgewell/trace-mapping@0.3.25", "", { "dependencies": { "@jridgewell/resolve-uri": "^3.1.0", "@jridgewell/sourcemap-codec": "^1.4.14" } }, "sha512-vNk6aEwybGtawWmy/PzwnGDOjCkLWSD2wqvjGGAgOAwCGWySYXfYoxt00IJkTF+8Lb57DwOb3Aa0o9CApepiYQ=="],

    "@standard-schema/spec": ["@standard-schema/spec@1.0.0", "", {}, "sha512-m2bOd0f2RT9k8QJx1JN85cZYyH1RqFBdlwtkSlf4tBDYLCiiZnv1fIIwacK6cqwXavOydf0NPToMQgpKq+dVlA=="],

    "@std/semver": ["@jsr/std__semver@1.0.4", "https://npm.jsr.io/~/11/@jsr/std__semver/1.0.4.tgz", {}, "sha512-QdKjWkoc/OSr+e5OkMX6jCuy9+QFU3xb3InOESUnrfUmI2iC0zbXxq7m2GNkvEAwpqmrSqanDhrgkqETgk8TFA=="],

    "@trivago/prettier-plugin-sort-imports": ["@trivago/prettier-plugin-sort-imports@5.2.2", "", { "dependencies": { "@babel/generator": "^7.26.5", "@babel/parser": "^7.26.7", "@babel/traverse": "^7.26.7", "@babel/types": "^7.26.7", "javascript-natural-sort": "^0.7.1", "lodash": "^4.17.21" }, "peerDependencies": { "@vue/compiler-sfc": "3.x", "prettier": "2.x - 3.x", "prettier-plugin-svelte": "3.x", "svelte": "4.x || 5.x" }, "optionalPeers": ["@vue/compiler-sfc", "prettier-plugin-svelte", "svelte"] }, "sha512-fYDQA9e6yTNmA13TLVSA+WMQRc5Bn/c0EUBditUHNfMMxN7M82c38b1kEggVE3pLpZ0FwkwJkUEKMiOi52JXFA=="],

    "@types/bun": ["@types/bun@1.2.4", "", { "dependencies": { "bun-types": "1.2.4" } }, "sha512-QtuV5OMR8/rdKJs213iwXDpfVvnskPXY/S0ZiFbsTjQZycuqPbMW8Gf/XhLfwE5njW8sxI2WjISURXPlHypMFA=="],

    "@types/node": ["@types/node@20.12.14", "", { "dependencies": { "undici-types": "~5.26.4" } }, "sha512-scnD59RpYD91xngrQQLGkE+6UrHUPzeKZWhhjBSa3HSkwjbQc38+q3RoIVEwxQGRw3M+j5hpNAM+lgV3cVormg=="],

    "@types/ws": ["@types/ws@8.5.12", "", { "dependencies": { "@types/node": "*" } }, "sha512-3tPRkv1EtkDpzlgyKyI8pGsGZAGPEaXeu0DOj5DI25Ja91bdAYddYHbADRYVrZMRbfW+1l5YwXVDKohDJNQxkQ=="],

    "ansi-regex": ["ansi-regex@5.0.1", "", {}, "sha512-quJQXlTSUGL2LH9SUXo8VwsY4soanhgo6LNSm84E1LBcE8s3O0wpdiRzyR9z/ZZJMlMWv37qOOb9pdJlMUEKFQ=="],

    "ansi-styles": ["ansi-styles@4.3.0", "", { "dependencies": { "color-convert": "^2.0.1" } }, "sha512-zbB9rCJAT1rbjiVDb2hqKFHNYLxgtk8NURxZ3IZwD3F6NtxbXZQCnnSi1Lkx+IDohdPlFp222wVALIheZJQSEg=="],

    "arktype": ["arktype@2.1.2", "", { "dependencies": { "@ark/schema": "0.43.1", "@ark/util": "0.43.1" } }, "sha512-+uwghWzB1C0WoN6TpVB6wqXpshbRu19LykgddVcD8fk2SZHSgXWcxaTnq+cHnAqtZZOfJ/OI/QDTBmf9SiUFCw=="],

    "bun-types": ["bun-types@1.2.4", "", { "dependencies": { "@types/node": "*", "@types/ws": "~8.5.10" } }, "sha512-nDPymR207ZZEoWD4AavvEaa/KZe/qlrbMSchqpQwovPZCKc7pwMoENjEtHgMKaAjJhy+x6vfqSBA1QU3bJgs0Q=="],

    "cbor2": ["cbor2@1.12.0", "", {}, "sha512-3Cco8XQhi27DogSp9Ri6LYNZLi/TBY/JVnDe+mj06NkBjW/ZYOtekaEU4wZ4xcRMNrFkDv8KNtOAqHyDfz3lYg=="],

    "cliui": ["cliui@8.0.1", "", { "dependencies": { "string-width": "^4.2.0", "strip-ansi": "^6.0.1", "wrap-ansi": "^7.0.0" } }, "sha512-BSeNnyus75C4//NQ9gQt1/csTXyo/8Sb+afLAkzAptFuMsod9HFokGNudZpi/oQV73hnVK+sR+5PVRMd+Dr7YQ=="],

    "color-convert": ["color-convert@2.0.1", "", { "dependencies": { "color-name": "~1.1.4" } }, "sha512-RRECPsj7iu/xb5oKYcsFHSppFNnsj/52OVTRKb4zP5onXwVF3zVmmToNcOfGC+CRDpfK/U584fMg38ZHCaElKQ=="],

    "color-name": ["color-name@1.1.4", "", {}, "sha512-dOy+3AuW3a2wNbZHIuMZpTcgjGuLU/uBL/ubcZF9OXbDo8ff4O8yVp5Bf0efS8uEoYo5q4Fx7dY9OgQGXgAsQA=="],

    "dayjs": ["dayjs@1.11.13", "", {}, "sha512-oaMBel6gjolK862uaPQOVTA7q3TZhuSvuMQAAglQDOWYO9A91IrAOUJEyKVlqJlHE0vq5p5UXxzdPfMH/x6xNg=="],

    "debug": ["debug@4.3.7", "", { "dependencies": { "ms": "^2.1.3" } }, "sha512-Er2nc/H7RrMXZBFCEim6TCmMk02Z8vLC2Rbi1KEBggpo0fS6l0S1nnapwmIi3yW/+GOJap1Krg4w0Hg80oCqgQ=="],

    "dts-bundle-generator": ["dts-bundle-generator@9.5.1", "", { "dependencies": { "typescript": ">=5.0.2", "yargs": "^17.6.0" }, "bin": { "dts-bundle-generator": "dist/bin/dts-bundle-generator.js" } }, "sha512-DxpJOb2FNnEyOzMkG11sxO2dmxPjthoVWxfKqWYJ/bI/rT1rvTMktF5EKjAYrRZu6Z6t3NhOUZ0sZ5ZXevOfbA=="],

    "emoji-regex": ["emoji-regex@8.0.0", "", {}, "sha512-MSjYzcWNOA0ewAHpz0MxpYFvwg6yjy1NG3xteoqz644VCo/RPgnr1/GGt+ic3iJTzQ8Eu3TdM14SawnVUmGE6A=="],

    "escalade": ["escalade@3.2.0", "", {}, "sha512-WUj2qlxaQtO4g6Pq5c29GTcWGDyd8itL8zTlipgECz3JesAiiOKotd8JU6otB3PACgG6xkJUyVhboMS+bje/jA=="],

    "get-caller-file": ["get-caller-file@2.0.5", "", {}, "sha512-DyFP3BM/3YHTQOCUL/w0OZHR0lpKeGrxotcHWcqNEdnltqFwXVfhEBQ94eIo34AfQpo0rGki4cyIiftY06h2Fg=="],

    "globals": ["globals@11.12.0", "", {}, "sha512-WOBp/EEGUiIsJSp7wcv/y6MO+lV9UoncWqxuFfm8eBwzWNgyfBd6Gz+IeKQ9jCmyhoH99g15M3T+QaVHFjizVA=="],

    "is-fullwidth-code-point": ["is-fullwidth-code-point@3.0.0", "", {}, "sha512-zymm5+u+sCsSWyD9qNaejV3DFvhCKclKdizYaJUuHA83RLjb7nSuGnddCHGv0hk+KY7BMAlsWeK4Ueg6EV6XQg=="],

    "javascript-natural-sort": ["javascript-natural-sort@0.7.1", "", {}, "sha512-nO6jcEfZWQXDhOiBtG2KvKyEptz7RVbpGP4vTD2hLBdmNQSsCiicO2Ioinv6UI4y9ukqnBpy+XZ9H6uLNgJTlw=="],

    "js-tokens": ["js-tokens@4.0.0", "", {}, "sha512-RdJUflcE3cUzKiMqQgsCu06FPu9UdIJO0beYbPhHN4k6apgJtifcoCtT9bcxOpYBtpD2kCM6Sbzg4CausW/PKQ=="],

    "jsesc": ["jsesc@3.0.2", "", { "bin": { "jsesc": "bin/jsesc" } }, "sha512-xKqzzWXDttJuOcawBt4KnKHHIf5oQ/Cxax+0PWFG+DFDgHNAdi+TXECADI+RYiFUMmx8792xsMbbgXj4CwnP4g=="],

    "lodash": ["lodash@4.17.21", "", {}, "sha512-v2kDEe57lecTulaDIuNTPy3Ry4gLGJ6Z1O3vE1krgXZNrsQ+LFTGHVxVjcXPs17LhbZVGedAJv8XZ1tvj5FvSg=="],

    "ms": ["ms@2.1.3", "", {}, "sha512-6FlzubTLZG3J2a/NVCAleEhjzq5oxgHyaCU9yYXvcLsvoVaHJq/s5xXI6/XXP6tz7R9xAOtHnSO/tXtF3WRTlA=="],

    "picocolors": ["picocolors@1.1.1", "", {}, "sha512-xceH2snhtb5M9liqDsmEw56le376mTZkEX/jEb/RxNFyegNul7eNslCXP9FDj/Lcu0X8KEyMceP2ntpaHrDEVA=="],

    "prettier": ["prettier@3.5.3", "", { "bin": { "prettier": "bin/prettier.cjs" } }, "sha512-QQtaxnoDJeAkDvDKWCLiwIXkTgRhwYDEQCghU9Z6q03iyek/rxRh/2lC3HB7P8sWT2xC/y5JDctPLBIGzHKbhw=="],

    "require-directory": ["require-directory@2.1.1", "", {}, "sha512-fGxEI7+wsG9xrvdjsrlmL22OMTTiHRwAMroiEeMgq8gzoLC/PQr7RsRDSTLUg/bZAZtF+TVIkHc6/4RIKrui+Q=="],

    "string-width": ["string-width@4.2.3", "", { "dependencies": { "emoji-regex": "^8.0.0", "is-fullwidth-code-point": "^3.0.0", "strip-ansi": "^6.0.1" } }, "sha512-wKyQRQpjJ0sIp62ErSZdGsjMJWsap5oRNihHhu6G7JVO/9jIB6UyevL+tXuOqrng8j/cxKTWyWUwvSTriiZz/g=="],

    "strip-ansi": ["strip-ansi@6.0.1", "", { "dependencies": { "ansi-regex": "^5.0.1" } }, "sha512-Y38VPSHcqkFrCpFnQ9vuSXmquuv5oXOKpGeT6aGrr3o3Gc9AlVa6JBfUSOCnbxGGZF+/0ooI7KrPuUSztUdU5A=="],

    "typescript": ["typescript@5.8.2", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-aJn6wq13/afZp/jT9QZmwEjDqqvSGp1VT5GVg+f/t6/oVyrgXM6BY1h9BRh/O5p3PlUPAe+WuiEZOmb/49RqoQ=="],

    "undici-types": ["undici-types@5.26.5", "", {}, "sha512-JlCMO+ehdEIKqlFxk6IfVoAUVmgz7cU7zD/h9XZ0qzeosSHmUJVOzSQvvYSYWXkFXC+IfLKSIffhv0sVZup6pA=="],

    "wrap-ansi": ["wrap-ansi@7.0.0", "", { "dependencies": { "ansi-styles": "^4.0.0", "string-width": "^4.1.0", "strip-ansi": "^6.0.0" } }, "sha512-YVGIj2kamLSTxw6NsZjoBxfSwsn0ycdesmc4p+Q21c5zPuZ1pl+NfxVdxPtdHvmNVOQ6XSYG4AUtyt/Fi7D16Q=="],

    "y18n": ["y18n@5.0.8", "", {}, "sha512-0pfFzegeDWJHJIAmTLRP2DwHjdF5s7jo9tuztdQxAhINCdvS+3nGINqPd00AphqJR/0LhANUS6/+7SCb98YOfA=="],

    "yargs": ["yargs@17.7.2", "", { "dependencies": { "cliui": "^8.0.1", "escalade": "^3.1.1", "get-caller-file": "^2.0.5", "require-directory": "^2.1.1", "string-width": "^4.2.3", "y18n": "^5.0.5", "yargs-parser": "^21.1.1" } }, "sha512-7dSzzRQ++CKnNI/krKnYRV7JKKPUXMEh61soaHKg9mrWEhzFWhFnxPxGl+69cD1Ou63C13NUPCnmIcrvqCuM6w=="],

    "yargs-parser": ["yargs-parser@21.1.1", "", {}, "sha512-tVpsJW7DdjecAiFpbIB1e3qxIQsE6NoPc5/eTdrbbIC4h0LVsWhnoa3g+m2HclBIujHzsxZ4VJVA+GUuc2/LBw=="],
  }
}
