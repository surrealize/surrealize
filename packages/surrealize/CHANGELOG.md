# surrealize

## 0.1.0-alpha.7

### Patch Changes

- [`7dc0b07`](https://github.com/surrealize/surrealize/commit/7dc0b07ffcf25b4eea134ca9c2552a9e10340504) Thanks [@lkwr](https://github.com/lkwr)! - Make "keep" types more strict

## 0.1.0-alpha.6

### Patch Changes

- [`4214a42`](https://github.com/surrealize/surrealize/commit/4214a425107514e75a0c86367a6eefe01feba928) Thanks [@lkwr](https://github.com/lkwr)! - Fix: ensured "keep" are not producing type errors

## 0.1.0-alpha.5

### Patch Changes

- [`72b316b`](https://github.com/surrealize/surrealize/commit/72b316b2e1da3d858143fb5d9975cc9847d2a934) Thanks [@lkwr](https://github.com/lkwr)! - Add the ability to skip flatting specific values

## 0.1.0-alpha.4

### Minor Changes

- [`8cb2b94`](https://github.com/surrealize/surrealize/commit/8cb2b94038df0575e584d65c3ba184cb288c1ec1) Thanks [@lkwr](https://github.com/lkwr)! - Added Repository "updateBy" and "updateById" methods

## 0.1.0-alpha.3

### Minor Changes

- [`7c0023f`](https://github.com/surrealize/surrealize/commit/7c0023f8c071f3c2158b4be9ebc5df76c7ac0652) Thanks [@lkwr](https://github.com/lkwr)! - Added "deleteBy" repository method

## 0.1.0-alpha.2

### Minor Changes

- [`aec8e4e`](https://github.com/surrealize/surrealize/commit/aec8e4eb271fff5f522e2cc3c9fba37569b82121) Thanks [@lkwr](https://github.com/lkwr)! - Added "fetch" clause to "select"

## 0.1.0-alpha.1

### Patch Changes

- [`058fd04`](https://github.com/surrealize/surrealize/commit/058fd04e1d2c37e6c499e864e4a53b3435e449a9) Thanks [@lkwr](https://github.com/lkwr)! - Export Repository APIs

## 0.1.0-alpha.0

### Minor Changes

- [`32d2c7e`](https://github.com/surrealize/surrealize/commit/32d2c7e10713c6c07a8119f319edbc0c149e4eb3) Thanks [@lkwr](https://github.com/lkwr)! - Initial release
