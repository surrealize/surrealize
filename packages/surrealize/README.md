# Surrealize

## TODO

- not compiling yet
