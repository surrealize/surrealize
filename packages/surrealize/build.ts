import { copyFile } from "node:fs/promises";
import { build } from "tsup";

import packageJson from "./package.json" with { type: "json" };

// bundle the library
await build({
	entry: [`${import.meta.dirname}/src/index.ts`],
	outDir: `${import.meta.dirname}/dist`,
	format: ["esm", "cjs"],
	dts: true,
	clean: true,
	minify: true,
	platform: "neutral",
	sourcemap: true,
});

// write package.json
await Bun.write(
	`${import.meta.dirname}/dist/package.json`,
	JSON.stringify(
		{
			name: packageJson.name,
			version: packageJson.version,
			type: packageJson.type,

			exports: {
				".": {
					types: "./index.d.ts",
					import: "./index.js",
					require: "./index.cjs",
				},
			},

			dependencies: packageJson.dependencies,
			license: packageJson.license,
			author: packageJson.author,
			homepage: packageJson.homepage,
			repository: packageJson.repository,
			bugs: packageJson.bugs,
			keywords: packageJson.keywords,
		},
		null,
		2,
	),
);

await Promise.all([
	copyFile(
		`${import.meta.dirname}/../../README.md`,
		`${import.meta.dirname}/dist/README.md`,
	),
]);

console.log("ðŸš€ Build complete");
