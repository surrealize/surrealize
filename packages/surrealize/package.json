{
	"name": "surrealize",
	"version": "0.1.0-alpha.8",
	"description": "Surrealize is a query builder and ORM for SurrealDB.",
	"type": "module",
	"scripts": {
		"build": "bun run ./build.ts"
	},
	"dependencies": {
		"surrealdb": "^1.1.0"
	},
	"devDependencies": {
		"@surrealdb/wasm": "^1.1.0",
		"@types/bun": "^1.1.14",
		"tsup": "^8.3.5",
		"typescript": "^5.7.2"
	},
	"exports": {
		".": "./src/index.ts"
	},
	"license": "MIT",
	"author": "Lukas Heizmann <lukas@heizmann.dev>",
	"homepage": "https://github.com/surrealize/surrealize#readme",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/surrealize/surrealize.git",
		"directory": "packages/surrealize"
	},
	"bugs": {
		"url": "https://github.com/surrealize/surrealize/issues"
	},
	"keywords": [
		"surrealdb",
		"typescript",
		"orm",
		"query",
		"builder",
		"database"
	]
}
