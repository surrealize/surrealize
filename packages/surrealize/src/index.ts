export { Statements, q } from "./statements.ts";
export { Surrealize, type SurrealizeOptions } from "./surrealize.ts";

export { DEFAULT_CODECS } from "./query/transformer/codecs.ts";
export {
	Transformer,
	type TransformCodec,
	type TransformerCustomType,
	type Decodeable,
	type Encodeable,
} from "./query/transformer/transformer.ts";

export {
	QueryBuilder,
	prepareQuery,
	prepareTransaction,
	resolveQuery,
} from "./query/builder.ts";
export {
	Query,
	QueryList,
	surql,
	type QueryOptions,
	type QueryListOptions,
} from "./query/query.ts";
export {
	type FormatVariableConverter,
	type TaggedTemplate,
	format,
	isEmpty,
	merge,
	tag,
	tagString,
} from "./query/template.ts";
export type {
	PreparedQuery,
	QueryLike,
	QueriesLike,
	InferQueryOutput,
	InferQueriesOutput,
} from "./query/types.ts";

export { Repository, type RepositoryOptions } from "./repository/repository.ts";
export type {
	RepositoryWhere,
	RepositoryFindOptions,
	RepositoryFindByOptions,
	RepositoryFindOneOptions,
	RepositoryFindOneByOptions,
} from "./repository/types.ts";

export * from "./statement/shared/where.ts";

export {
	RecordId,
	type RecordIdLike,
	type RecordIdValue,
} from "./type/recordid.ts";
export { Table, type TableLike } from "./type/table.ts";
export {
	resolveTarget,
	type TargetLike,
	type ResolvedTarget,
} from "./type/target.ts";

export {
	type SchemaFunction,
	type SchemaLike,
	type InferSchemaOutput,
} from "./schema/types.ts";
export { mergeSchema, parseSchema } from "./schema/utils.ts";
