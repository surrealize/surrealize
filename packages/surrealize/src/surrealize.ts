import { Connection } from "./connection/connection.ts";
import type { EngineInitializer } from "./connection/engine.ts";
import { surql } from "./query/query.ts";
import type {
	InferQueriesOutput,
	QueriesLike,
	QueryLike,
} from "./query/types.ts";
import {
	prepareQuery,
	prepareTransaction,
	resolveQuery,
} from "./query/utils.ts";
import type { Schema } from "./schema/types.ts";
import { parseSchema } from "./schema/utils.ts";
import { RecordId } from "./type/recordid.ts";
import { type TargetLike, resolveTarget } from "./type/target.ts";

export type SurrealizeOptions = {
	url: URL | string;

	/**
	 * Set the connection as the default connection.
	 *
	 * This will make the Surrealize instance the default instance for all queries.
	 */
	default?: boolean;

	/**
	 * A collection of engines which are available to connect to.
	 *
	 * Defaults to the following engines:
	 * - `ws`: WebSocket
	 * - `wss`: WebSocket Secure
	 */
	engines?: Record<string, EngineInitializer>;
};

export class Surrealize {
	static default: Surrealize | undefined = undefined;

	/**
	 * The underlying SurrealDB connection to execute queries.
	 */
	connection: Connection;

	constructor(options: SurrealizeOptions) {
		this.connection = new Connection(
			{ url: options.url instanceof URL ? options.url : new URL(options.url) },
			{ engines: options.engines },
		);

		if (options.default) Surrealize.default = this;
	}

	async execute<TSchemaOutput>(
		queryLike: QueryLike<TSchemaOutput>,
	): Promise<TSchemaOutput> {
		const { template, schema } = resolveQuery(queryLike);
		const { query, bindings } = prepareQuery(template);

		const [result] = await this.connection.query(query, bindings);

		return (schema ? parseSchema(schema, result) : result) as TSchemaOutput;
	}

	async executeAll<const TQueries extends QueriesLike>(
		queriesLike: TQueries,
	): Promise<InferQueriesOutput<TQueries>> {
		return Promise.all(
			queriesLike.map((queryLike) =>
				this.execute(queryLike),
			) as InferQueriesOutput<TQueries>,
		);
	}

	async executeTransaction<const TQueries extends QueriesLike>(
		queriesLike: TQueries,
	): Promise<InferQueriesOutput<TQueries>> {
		const queries = queriesLike.map((queryLike) => resolveQuery(queryLike));

		const { query, bindings } = prepareTransaction(queries);

		const results = await this.connection.query(query, bindings);

		return queries.map(({ schema }, index) => {
			const queryResult = results[index];
			return schema ? parseSchema(schema, queryResult) : queryResult;
		}) as InferQueriesOutput<TQueries>;
	}

	/**
	 * Get a specific target from the database. A target can be a record id or a table.
	 * In case the target is a table, the result will be an array of records.
	 *
	 * @param targetLike A target like input (record id or table).
	 * @param schema The optional schema to use for validating the result.
	 * @returns The target from the database.
	 */
	async resolve<TOutput>(
		targetLike: TargetLike,
		schema?: Schema<TOutput>,
	): Promise<TOutput> {
		const target = resolveTarget(targetLike);

		const query =
			target instanceof RecordId
				? surql`SELECT * FROM ONLY ${target}`
				: surql`SELECT * FROM ${target}`;

		return this.execute(query.withSchema(schema));
	}
}
