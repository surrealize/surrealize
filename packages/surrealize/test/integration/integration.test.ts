import { surrealdbWasmEngines } from "@surrealdb/wasm";
import { beforeEach, describe, expect, test } from "bun:test";
import Surreal from "surrealdb";
import { Surrealize, surql } from "surrealize";

import { testFilter } from "./jobs/filter.ts";
import { testVersion } from "./jobs/version.ts";

const surrealize: Surrealize = new Surrealize({
	surreal: new Surreal({ engines: surrealdbWasmEngines() }),
});

beforeEach(async () => {
	await surrealize.connection.connect("mem:");
	await surrealize.connection.use({ namespace: "test", database: "test" });
});

describe("integration", () => {
	test("connection", async () => {
		const random = Math.floor(Math.random() * 100);
		const result = await surrealize.execute(surql`RETURN ${random}`);

		expect(result).toEqual(random);
	});

	test("version", () => testVersion(surrealize));
	test("filter", () => testFilter(surrealize));
});
