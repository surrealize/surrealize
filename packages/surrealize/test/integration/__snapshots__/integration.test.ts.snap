// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration filter 1`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 10,
    "id": RecordId {
      "table": "user",
      "value": "bob",
    },
    "name": "Bob",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;

exports[`Integration filter 2`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
]
`;

exports[`Integration filter 3`] = `
[
  {
    "age": 25,
    "id": RecordId {
      "table": "user",
      "value": "alice",
    },
    "name": "Alice",
  },
  {
    "age": 65,
    "id": RecordId {
      "table": "user",
      "value": "charlie",
    },
    "name": "Charlie",
  },
]
`;
